token-metadata/program/src/state.rs

```diff
@@ -89,11 +89,16 @@ pub const USE_AUTHORITY_RECORD_SIZE: usize = 18; //8 byte padding
 
 pub const COLLECTION_AUTHORITY_RECORD_SIZE: usize = 11; //10 byte padding
 
-pub trait TokenMetadataAccount {
+pub trait TokenMetadataAccount: BorshDeserialize {
     fn key() -> Key;
 
     fn size() -> usize;
 
+    fn is_correct_account_type(data: &[u8], data_type: Key, data_size: usize) -> bool {
+        (data[0] == data_type as u8 || data[0] == Key::Uninitialized as u8)
```

Comment:
As a general mention, I think the `as` operator should be avoided whenever possible; contrary to the core philosophy of the language, it quietly allows all sort of potentially dangerous casts, including (but not limited) to stuff like lossy narrowing of types. While a bit more verbose, I think it's better to perform conversions via specific methods (or implementations of traits like `From/Into`; or fallible stuff like `try_...` which return results if it's possible in any way to have failures).